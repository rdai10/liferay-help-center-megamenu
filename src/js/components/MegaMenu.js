import preact from 'preact';
import PropTypes from 'prop-types';

import Card from './Card';
import CardMenu from './CardMenu';

class Submenu extends preact.Component {
	constructor(props) {
		super(props);

		this.handleClick = this.handleClick.bind(this);

		this.state = {
			selectLanguage: false
		};
	}

	handleClick() {
		this.setState(
			{
				selectLanguage: !this.state.selectLanguage
			}
		);
	}

	render({accountLinks, highlightedLinks, locale}, {selectLanguage}) {
		return selectLanguage ? (
			<div class="col-3 language-selection submenu">
				<svg class="icon" onClick={this.handleClick}>
					<use xlinkHref="#arrow-left" />
				</svg>

				<div class="available-languages">
					<a
						class="current"
						href={locale.currentLocale.url}
						rel="nofollow"
					>
						{locale.currentLocale.name}
					</a>

					{locale.alternativeLocales.map(
						alternativeLocale => (
							<a href={alternativeLocale.url} rel="nofollow">
								{alternativeLocale.name}
							</a>
						)
					)}
				</div>
			</div>
		) : (
			<div class="col-3 submenu">
				<CardMenu
					className={highlightedLinks.className}
					configs={highlightedLinks.configs}
					type="nav"
				/>

				<CardMenu
					className={accountLinks.className}
					configs={accountLinks.configs}
					type="nav"
				/>

				<div class="language">
					<Card
						name={locale.currentLocale.name}
						svgId="#language"
						type="nav"
						onClick={this.handleClick}
						url="javascript:;"
					/>
				</div>
			</div>
		);
	}
}

Submenu.propTypes = {
	accountLinks: PropTypes.object,
	highlightedLinks: PropTypes.object,
	locale: PropTypes.object
};

const MegaMenu = (
	{
		accountLinks,
		cardMenuItems,
		highlightedLinks,
		locale,
		name
	}
) => (
	<div class="container-fluid container-fluid-max-xl">
		<div class="header-menu-content row">
			<div class="col-9 menu-body">
				<h6 class="secondary-text-color">
					{name}
				</h6>

				<div class="card-menu-container">
					<CardMenu
						className={cardMenuItems.className}
						configs={cardMenuItems.configs}
						type="nav"
					/>
				</div>
			</div>

			<Submenu
				accountLinks={accountLinks}
				highlightedLinks={highlightedLinks}
				locale={locale}
			/>
		</div>
	</div>
);

MegaMenu.propTypes = {
	accountLinks: PropTypes.object,
	cardMenuItems: PropTypes.object,
	highlightedLinks: PropTypes.object,
	locale: PropTypes.object,
	name: PropTypes.string
};

export default MegaMenu;