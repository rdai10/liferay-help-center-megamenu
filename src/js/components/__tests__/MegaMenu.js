import preact from 'preact';
import {
	cleanup,
	debounceRenderingOff,
	fireEvent,
	render,
	queryByRole
} from 'preact-testing-library';

import MegaMenu from '../MegaMenu';

const setup = () => {
	const configs = [
		{
			name: 'Item 1'
		},
		{
			svgId: '#glasses',
			url: '/'
		},
		{
			name: 'Item 3',
			svgId: '#sushi',
			url: '/'
		}
	];

	const locale = {
		alternativeLocales: [
			{
				name: 'fr',
				url: '/'
			},
			{
				name: 'de',
				url: '/'
			}
		],
		currentLocale: {
			name: 'en',
			url: '/'
		}
	};

	const utils = render(
		<MegaMenu
			accountLinks={{configs: configs}}
			cardMenuItems={{configs: configs}}
			highlightedLinks={{configs: configs}}
			locale={locale}
			name="Test"
		/>
	);

	return {...utils};
};

afterEach(cleanup);

describe('MegaMenu', () => {
	it('renders menu correctly', () => {
		const {container} = setup();

		expect(container).toMatchSnapshot();
	});

	it('renders a list of language selectors on clicking the current language', () => {
		debounceRenderingOff();

		const {container, getByText, queryByText} = setup();

		expect(queryByText('en')).toBeTruthy();

		fireEvent.click(getByText('en'));

		expect(queryByText('fr')).toBeTruthy();
		expect(queryByText('de')).toBeTruthy();
		expect(container).toMatchSnapshot();
	});

	it('returns to main mega menu when clicking on the back arrow in language selection', () => {
		const {getByText, queryByRole, queryByText} = setup();

		fireEvent.click(getByText('en'));

		expect(queryByText('fr')).toBeTruthy();
		expect(queryByText('de')).toBeTruthy();

		fireEvent.click(queryByRole('img'));

		expect(queryByText('fr')).toBeFalsy();
		expect(queryByText('de')).toBeFalsy();
	});
});