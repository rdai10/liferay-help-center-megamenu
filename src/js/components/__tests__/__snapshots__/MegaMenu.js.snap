// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MegaMenu renders a list of language selectors on clicking the current language 1`] = `
<div>
  <div
    class="container-fluid container-fluid-max-xl"
  >
    <div
      class="header-menu-content row"
    >
      <div
        class="col-md-9 menu-body"
      >
        <h6
          class="secondary-text-color"
        >
          Test
        </h6>
        <div
          class="card-menu-container"
        >
          <div
            class=""
          >
            <div
              class="nav-card"
            >
              <a
                class="autofit-row autofit-row-center"
              >
                
                <div
                  class="autofit-col autofit-col-expand"
                >
                  <h4
                    class="title"
                  >
                    Item 1
                  </h4>
                  
                </div>
              </a>
            </div>
            <div
              class="nav-card"
            >
              <a
                class="autofit-row autofit-row-center"
                href="/"
              >
                <div
                  class="autofit-col"
                >
                  <svg
                    class="icon"
                  >
                    <use
                      href="#glasses"
                    />
                  </svg>
                </div>
                <div
                  class="autofit-col autofit-col-expand"
                >
                  
                </div>
              </a>
            </div>
            <div
              class="nav-card"
            >
              <a
                class="autofit-row autofit-row-center"
                href="/"
              >
                <div
                  class="autofit-col"
                >
                  <svg
                    class="icon"
                  >
                    <use
                      href="#sushi"
                    />
                  </svg>
                </div>
                <div
                  class="autofit-col autofit-col-expand"
                >
                  <h4
                    class="title"
                  >
                    Item 3
                  </h4>
                  
                </div>
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-md-3 language-selection submenu"
        id="languageToggle"
      >
        <svg
          aria-describedby="back to language selection"
          aria-labelledby="back-button"
          class="icon"
          role="img"
        >
          <use
            href="#arrow-left"
          />
        </svg>
        <div
          class="available-languages"
        >
          <a
            class="current"
            href="/"
            rel="nofollow"
          >
            en
          </a>
          <a
            href="/"
            rel="nofollow"
          >
            fr
          </a>
          <a
            href="/"
            rel="nofollow"
          >
            de
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MegaMenu renders menu correctly 1`] = `
<div>
  <div
    class="container-fluid container-fluid-max-xl"
  >
    <div
      class="header-menu-content row"
    >
      <div
        class="col-md-9 menu-body"
      >
        <h6
          class="secondary-text-color"
        >
          Test
        </h6>
        <div
          class="card-menu-container"
        >
          <div
            class=""
          >
            <div
              class="nav-card"
            >
              <a
                class="autofit-row autofit-row-center"
              >
                
                <div
                  class="autofit-col autofit-col-expand"
                >
                  <h4
                    class="title"
                  >
                    Item 1
                  </h4>
                  
                </div>
              </a>
            </div>
            <div
              class="nav-card"
            >
              <a
                class="autofit-row autofit-row-center"
                href="/"
              >
                <div
                  class="autofit-col"
                >
                  <svg
                    class="icon"
                  >
                    <use
                      href="#glasses"
                    />
                  </svg>
                </div>
                <div
                  class="autofit-col autofit-col-expand"
                >
                  
                </div>
              </a>
            </div>
            <div
              class="nav-card"
            >
              <a
                class="autofit-row autofit-row-center"
                href="/"
              >
                <div
                  class="autofit-col"
                >
                  <svg
                    class="icon"
                  >
                    <use
                      href="#sushi"
                    />
                  </svg>
                </div>
                <div
                  class="autofit-col autofit-col-expand"
                >
                  <h4
                    class="title"
                  >
                    Item 3
                  </h4>
                  
                </div>
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-md-3 submenu"
      >
        <div
          class=""
        >
          <div
            class="nav-card"
          >
            <a
              class="autofit-row autofit-row-center"
            >
              
              <div
                class="autofit-col autofit-col-expand"
              >
                <h4
                  class="title"
                >
                  Item 1
                </h4>
                
              </div>
            </a>
          </div>
          <div
            class="nav-card"
          >
            <a
              class="autofit-row autofit-row-center"
              href="/"
            >
              <div
                class="autofit-col"
              >
                <svg
                  class="icon"
                >
                  <use
                    href="#glasses"
                  />
                </svg>
              </div>
              <div
                class="autofit-col autofit-col-expand"
              >
                
              </div>
            </a>
          </div>
          <div
            class="nav-card"
          >
            <a
              class="autofit-row autofit-row-center"
              href="/"
            >
              <div
                class="autofit-col"
              >
                <svg
                  class="icon"
                >
                  <use
                    href="#sushi"
                  />
                </svg>
              </div>
              <div
                class="autofit-col autofit-col-expand"
              >
                <h4
                  class="title"
                >
                  Item 3
                </h4>
                
              </div>
            </a>
          </div>
        </div>
        <div
          class="language"
        >
          <div
            class="nav-card"
            id="languageToggle"
          >
            <a
              class="autofit-row autofit-row-center"
              href="javascript:;"
            >
              <div
                class="autofit-col"
              >
                <svg
                  class="icon"
                >
                  <use
                    href="#language"
                  />
                </svg>
              </div>
              <div
                class="autofit-col autofit-col-expand"
              >
                <h4
                  class="title"
                >
                  en
                </h4>
                
              </div>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
